ENH: Handle pandas DataFrame with integer column names in ColumnTransformer

Tunahanyrd opened the pull request:
#### Reference Issues/PRs

Fixes #30983

#### What does this implement/fix? Explain your changes.

This PR makes `ColumnTransformer` raise a clear error when the input `DataFrame` has integer column names.
Integer column names are ambiguous in scikit-learn, because integers are interpreted as column *positions*.
Before this PR, this ambiguity could lead to confusing `ValueError`s.
Now, users will see a clear error message instructing them to cast column names to strings, e.g.:

```python
df.columns = df.columns.astype(str)
```

#### Any other comments?

This is one possible resolution of the ambiguity (explicit error).
An alternative could be autocasting + `FutureWarning`, but that would silently change user data.
Therefore, raising an error seems safer.
Happy to adjust if maintainers prefer the warning approach.

Tunahanyrd committed:
Raise error for integer column names in ColumnTransformer

Tunahanyrd committed:
Raise error for integer column names in ColumnTransformer

Tunahanyrd committed:
Raise error for integer column names in ColumnTransformer

adrinjalali reviewed (commented):
Instead of this, I think we should warn that feature names are not strings, and that we'd be ingoring them, and if they're integers, we ignore feature names and index based on the column position

WDYT @thomasjpfan @lesteve

Tunahanyrd committed:
ENH Warn on integer column names in _get_feature_names (FutureWarning)

thomasjpfan commented:
> Instead of this, I think we should warn that feature names are not strings, and that we'd be ingoring them, and if they're integers, we ignore feature names and index based on the column position

I'm okay with this since it's the most consistent way to handle it, since we always treat integers as positions.

Note, it will not resolve the issue from https://github.com/scikit-learn/scikit-learn/issues/30983#issue-2914104945 where they wanted to treat the integers as actual names.

lesteve commented:
The CI is red, unfortunately the logs are no longer available. Can you trigger the CI it e.g. by adding an empty commit and push it? 

You then will need to understand the failures and fix them.

Tunahanyrd committed:
ENH: Use UserWarning for ambiguous integer column names (fixes #30983)

Tunahanyrd committed:
ENH: Use UserWarning for ambiguous integer column names (fixes #30983)

Aligns the handling of pandas DataFrames with integer column names to the
maintainers' final decision by using a standard UserWarning.

- Removes the redundant ValueError check from ColumnTransformer.
- Updates the check in _get_feature_names to issue a UserWarning, clarifying
  that integer feature names are ignored and selection defaults to position.
- Updates affected tests in test_validation.py and test_base.py to correctly
  expect UserWarning instead of FutureWarning or no warning.

Fixes #30983

Tunahanyrd committed:
FIX: Suppress new UserWarning in estimator checks test

Addresses the failing 'test_xfail_count_with_no_fast_fail' by explicitly
ignoring the new UserWarning (regarding integer column names) that was
triggered during the check_estimator call. This restores the test's ability
to correctly count EstimatorCheckFailedWarnings.

adrinjalali commented on sklearn/tests/test_base.py:
```diff
@@ -777,17 +777,20 @@ def transform(self, X):
     # fit on dataframe with all integer feature names works without warning
```
comment needs amending

adrinjalali reviewed (commented):
another round of review

Tunahanyrd commented:
@adrinjalali , the only remaining blocker is the cython-lint issue in _tree.pyx (unnecessary set + generator).
Since this change is required for the CI to pass but falls outside the scope of integer column ambiguity, should I merge the necessary MAINT fix to clear the checks, or do you have a preferred method to temporarily suppress this specific linter rule for the PR?

adrinjalali reviewed (approved):
LGTM. Maybe @lorentzenchr wants to have a look as a second reviewer.

thomasjpfan commented on sklearn/utils/tests/test_estimator_checks.py:
```diff
@@ -1422,7 +1422,9 @@ def test_xfail_count_with_no_fast_fail():
         )
-    xfail_warns = [w for w in records if w.category != SkipTestWarning]
+    xfail_warns = [
+        w for w in records if w.category not in (SkipTestWarning, UserWarning)
```
I prefer to update our common tests to use DataFrames with string columns and not trigger a `UserWarning`.

Tunahanyrd committed:
for ci

lorentzenchr commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -1952,14 +1952,26 @@ def test_num_features_errors_scalars(X):
-def test_get_feature_names_pandas_with_ints_no_warning(names):
-    """Get feature names with pandas dataframes without warning.
+def test_get_feature_names_pandas_with_ints_warns(names):
+    """Integer column names (including default None) should warn."""
```
```suggestion
    """Test that column names of integer dtype raise warning."""
```

lorentzenchr commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -1952,14 +1952,26 @@ def test_num_features_errors_scalars(X):
     "names",
-    [list(range(2)), range(2), None, [["a", "b"], ["c", "d"]]],
-    ids=["list-int", "range", "default", "MultiIndex"],
+    [list(range(2)), range(2), None],
```
Could you add numpy array if integers, e.g. np.arange(2)?
I have a suspicion…

Tunahanyrd commented:
​Thank you very much for your help, but I am too busy and I don't think I can continue with this. If you would like, please continue from here so everything doesn't go to waste now that we've reached this stage.

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/Tunahanyrd/32170

jeremiedbb committed:
address review comments
